### PR TITLE
add support for ConnectionFactory ProviderFn in JmsIO

### PR DESCRIPTION
**Pull request by tilgalas**
addresses #18110

Added a new configuration method `withConnectionFactoryProviderFn` to the JmsIO's Read and Write transforms, which enables the user to specify a lazily instantiated CF objects in case they're not serializable

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java [hunk 1]
```diff
@@ -275,9 +306,33 @@ abstract static class Builder<T> {
      * @param connectionFactory The JMS {@link ConnectionFactory}.
      * @return The corresponding {@link JmsIO.Read}.
      */
+    @Override
```
Comment (Abacn) [inline sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java:309]: shall we note in javadoc that "myConnectionFactory" has to be serializable? If not then need to use withConnectionFactoryProviderFn

File: sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOIT.java [hunk 2]
```diff
@@ -156,12 +165,14 @@ public JmsIOIT(
       String brokerUrl,
       Integer brokerPort,
       String forceAsyncAcksParam,
-      Class<? extends ConnectionFactory> connectionFactoryClass) {
+      Class<? extends ConnectionFactory> connectionFactoryClass,
+      boolean useProviderFn) {
```
Comment (Abacn) [inline sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOIT.java:169]: Would you mind sync the order of constructor parameter with CommonJms ? that is swap connectionFactoryClass / useProviderFn

File: sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOIT.java [hunk 3]
```diff
@@ -232,12 +243,13 @@ private PipelineResult readMessages() {
     pipelineRead
         .apply(
             "Read Messages",
-            JmsIO.<String>readMessage()
+            commonJms
```
Comment (Abacn) [inline sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOIT.java:246]: our integration test is also served as example that how a specific IO can be used. Could you please either keep `.withConnectionFactory(connectionFactory)` or just use withConnectionFactoryFn(...).

Currently it sets "withConnectionFactory" but serving a connectionFactoryProviderFn though its get method, sounds a little bit unnatural.

The point is the configuration of JmsIO.<String>readMessage() can be served as an example thus minimize test-taste settings.

File: sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOTest.java [hunk 4]
```diff
@@ -124,46 +127,56 @@ public class JmsIOTest {
...
-  @Parameterized.Parameters(name = "with client class {3}")
+  private static Stream<List<Object>> connectionFactoriesParams() {
+    return Stream.of(
+        Arrays.asList(
+            "vm://localhost", 5672, "jms.sendAcksAsync=false", ActiveMQConnectionFactory.class),
+        Arrays.asList(
+            "amqp://localhost", 5672, "jms.forceAsyncAcks=false", JmsConnectionFactory.class));
+  }
+
+  @Parameterized.Parameters(name = "with client class {3}, use provider fn? {4}")
```
Comment (Abacn) [inline sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOTest.java:138]: will this double the number of tests?

We won't simply 2x number of test, which might become infeasible as number of test parameter increases, but will just add one more test case with this parameter changed.

I think we can either add one test case to ActiveMQConnectionFactory, or add one or a few tests testing connectionFactoryFn, while keep most existing tests untouched

File: sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOIT.java (hunk 2 above)
Comment (tilgalas) [reply sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOIT.java:169]: I guess there's a chance I won't be needing this parameter anymore when I apply your suggestion about not doubling the number of tests

**Comment by tilgalas** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by tilgalas** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by Abacn** [comment]
Java Examples Dataflow PreCommit currently failing on master branch: #31421 not related to the change

**Review by Abacn — APPROVED** [review]
Thank you!

**Pull request merged**